### PR TITLE
fix: Sort unordered set elements before committing to git repo so as to have same serialised output string for identical set of elements

### LINKED ISSUE #10932: [Bug]: Convert unordered Set to ordered one before committing the file to repo
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

As the `layoutOnLoadAction` have `List<Set<DslActionDTO>>` datatype which includes the unordered set, this shows unnecessary file changes when we serialize the objects to the JSON files. 

### Steps To Reproduce

1. Connect application to git
2. Create multiple `onPageLoadAction`
3. Try changes on canvas like widget displacement, changing the size of a widget which have nothing to do with the binding 
4. Check the committed files on remote repo where we can observe the changes in `layoutOnLoadActions`

### Environment

Release

### Version

Cloud

### PR DESCRIPTION
**Pull request by abhvsn**
## Description

> As Set is an unordered datatype, this is causing unnecessary changes to the file system during the commit operation even if the user has not altered anything since the last commit.

Fixes #10932

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

> JUnit TC
> Manual test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-git/src/main/java/com/appsmith/git/converters/GsonUnorderedToOrderedSetConverter.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+public class GsonUnorderedToOrderedSetConverter implements JsonSerializer<Set> {
+    @Override
+    public JsonArray serialize(Set src, Type typeOfSrc, JsonSerializationContext context) {
+        // Sort the set so that same elements will not end up in merge conflicts
+        return (JsonArray) new Gson().toJsonTree(getOrderedResource(src));
+    }
+
+    /**
+     * Sort the primitive datatype objects and string so that we will have predictable output
+     * @param objects set of objects which needs to be sorted before serialisation
```
Comment (AnaghHegde) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/converters/GsonUnorderedToOrderedSetConverter.java:24]: I felt the comment is little confusing. Can we add an example in the end of the current line?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CompareDslActionDTO.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.Comparator;
+
+public class CompareDslActionDTO implements Comparator<DslActionDTO>, Serializable {
+    // Method to compare DslActionDTO based on id
+    @Override
+    public int compare(DslActionDTO action1, DslActionDTO action2)
+    {
+        return action1.getId().compareTo(action2.getId());
+    }
+}
```
Comment (AnaghHegde) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CompareDslActionDTO.java:15]: Would it not work if we override the `comaprator` in DslActionDTO class ?

Comment (AnaghHegde) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CompareDslActionDTO.java:15]: Would it not work if we override the `comaprator` in DslActionDTO class ?

**Review by AnaghHegde — COMMENTED** [review]
Overall looks good to me other than the small nits. 
On a side thought, we are using the TreeSet for actions in a sorted manner. But being the TreeSet implementation asynchronous, wouldn't it cause any issues when there are multiple users trying to do git commit and status at the same time ?

**Comment by abhvsn** [comment]
> Overall looks good to me other than the small nits. On a side thought, we are using the TreeSet for actions in a sorted manner. But being the TreeSet implementation asynchronous, wouldn't it cause any issues when there are multiple users trying to do git commit and status at the same time?

TreeSet does not have reactive asynchronous implementation but it's not thread-safe. But I suppose we are working with the single thread here

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CompareDslActionDTO.java (hunk 2 above)
Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/CompareDslActionDTO.java:19]: I tried this but was getting following error 
```
java.lang.ClassCastException: class com.appsmith.server.dtos.DslActionDTO cannot be cast to class java.lang.Comparable (com.appsmith.server.dtos.DslActionDTO is in unnamed module of loader 'app'; java.lang.Comparable is in module java.base of loader 'bootstrap')
```

**Comment by abhvsn** [comment]
> Overall looks good to me other than the small nits. On a side thought, we are using the TreeSet for actions in a sorted manner. But being the TreeSet implementation asynchronous, wouldn't it cause any issues when there are multiple users trying to do git commit and status at the same time ?

TreeSet does not have reactive asynchronous implementation but it’s not thread-safe. But I suppose we are working with the single thread here, so it’s safe to assume this will work as expected. The other point you mentioned if multiple users make a call to commit, getStatus etc. `JGIT` handles the file locks but this can be improvised by taking control of file locks, nonetheless, this implementation doesn’t concern the above-mentioned issue.

**Review by AnaghHegde — APPROVED** [review]
LGTM

File: app/server/appsmith-git/src/main/java/com/appsmith/git/converters/GsonUnorderedToOrderedSetConverter.java (hunk 1 above)
Comment (abhvsn) [reply app/server/appsmith-git/src/main/java/com/appsmith/git/converters/GsonUnorderedToOrderedSetConverter.java:27]: Provided a sample example for ease of understanding
understanding

**Pull request merged**